src/path-decomposition.cpp

```diff
@@ -0,0 +1,167 @@
+        case VERTICES:
+            return "Not all vertices are in path-width decomposition";
+        default:
+            return "";
+    }
+}
+
+void PathDecomposition::Check() {
+    int n = boost::num_vertices(_g);
+
+    for (std::vector<vertex_t> &bag : _bags) {
+        std::sort(bag.begin(), bag.end());
+        bag.resize(std::unique(bag.begin(), bag.end()) - bag.begin());
+        //Erasing non-unique vertices from bags
+        for (int v : bag) {
+            if (v < 0 || v >= n)
+                throw PathDecomposition::CorectnessException(NEGATIVE_VERTICES);
+        }
+    }
+
+    std::vector<int> sum(n);
+    std::vector<int> left(n, n + 1);
+    std::vector<int> right(n, -1);
+
+    for (int i = 0; i < _bags.size(); ++i) {
+        for (vertex_t v : _bags[i]) {
+            sum[v]++;
+            left[v] = std::min(left[v], i);
+            right[v] = std::max(right[v], i);
+        }
+    }
+
+    for (int i = 0; i < n; ++i) {
+        if (sum[i] == 0) throw PathDecomposition::CorectnessException(VERTICES);
+        if (right[i] - left[i] + 1 != sum[i]) {
+            int j = left[i];
+            for (;
+                    std::find(_bags[j].begin(), _bags[j].end(), i) != _bags[j].end();
+                    ++j);
+            throw PathDecomposition::CorectnessException(left[i], j, right[i], i);
+        }
+    }
+
+    Graph path_g(n);
+
+    for (std::vector<vertex_t> bag : _bags) {
+        for (vertex_t v : bag) {
+            for (vertex_t u : bag) {
+                if (u == v) continue;
+                boost::add_edge(u, v, path_g);
+            }
+        }
+    }
+
+    for (int v = 0; v < n; ++v) {
+        for (auto it = boost::adjacent_vertices(v, _g).first; it != boost::adjacent_vertices(v, _g).second; ++it) {
+            if (std::find(boost::adjacent_vertices(v, path_g).first, boost::adjacent_vertices(v, path_g).second, *it) ==
+                boost::adjacent_vertices(v, path_g).second) throw PathDecomposition::CorectnessException(EDGES);
+        }
+    }
+}
+
+PathDecomposition::PathDecomposition(std::vector<std::vector<vertex_t>> bags, Graph g): _bags(bags), _g(g) {
+    Check();
+}
+
+PathDecomposition::PathDecomposition(Graph g): _g(g) {
+    transform();
+    Check();
+}
+
+void PathDecomposition::transform() {
+    Tree t;
+
+    treedec::exact_decomposition_cutset(_g, t);
+
+    auto tree_bags = get(treedec::bag_t(), t);
+    int log_n = 0, n = boost::num_vertices(t);
+    while ((1 << log_n) < n) ++log_n;
+    std::vector<std::vector<int>> size(n, std::vector<int>(log_n + 2));
+    // size[v][d] = size of subtree of vertex |v| on depth |d|
+    std::vector<int> used(n);
+    // used[v] = 1 if vertex |v| is already in centroid tree
+    // used[v] = 0 else
+    std::vector<int> parent(n, -1);
+    // parent[v] = parent of vertex |v| in centroid tree
+
+    std::function<void(int, int, int)> size_calculation;
+    size_calculation = [&](tree_vertex v, int depth, tree_vertex p) {
```

Comment:
I think this is too complicated to be inside a method. This makes it untestable separately which is not ideal given how error-prone tdlib is. Make a separate function that accepts a tree decomposition and returns a path decomposition.

Reply:
I think that never-used function will clutter up path-decomposition.h interface

Reply:
given the issues of tdlib, the testability of this algo separately is more important.